algebird-core/src/main/scala/com/twitter/algebird/maps/increment.scala

```diff
@@ -0,0 +1,159 @@
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird.maps.increment
+
+import math.Ordering
+
+import scala.collection.SortedMap
+
+import com.twitter.algebird.Monoid
+
+import com.twitter.algebird.maps.redblack.tree._
+import com.twitter.algebird.maps.ordered._
+import com.twitter.algebird.maps.ordered.tree.DataMap
+
+package tree {
+  import com.twitter.algebird.maps.ordered.tree._
+
+  /** Base trait of R/B tree nodes supporting increment */
+  trait NodeInc[K, V] extends NodeMap[K, V] {
+    /** The monoid that defines what it means to increment a value */
+    def valueMonoid: Monoid[V]
+
+    /** Increment the value at a key, by another value */
+    private[increment] final def incr(di: DataMap[K, V]) = blacken(inc(di))
+
+    private[tree] def inc(di: DataMap[K, V]): Node[K]
+  }
+
+  /** Leaf R/B node supporting increment */
+  trait LNodeInc[K, V] extends NodeInc[K, V] with LNodeMap[K, V] {
+    final def inc(di: DataMap[K, V]) = {
+      val d = new DataMap[K, V] {
+        val key = di.key
+        val value = valueMonoid.plus(valueMonoid.zero, di.value)
+      }
+      rNode(d, this, this)
+    }
+  }
+
+  /** Internal R/B node supporting increment */
+  trait INodeInc[K, V] extends NodeInc[K, V] with INodeMap[K, V] {
+    def lsub: NodeInc[K, V]
+    def rsub: NodeInc[K, V]
+
+    final def inc(di: DataMap[K, V]) =
+      if (color == R) {
+        if (keyOrdering.lt(di.key, data.key)) rNode(data, lsub.inc(di), rsub)
+        else if (keyOrdering.gt(di.key, data.key)) rNode(data, lsub, rsub.inc(di))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          rNode(d, lsub, rsub)
+        }
+      } else {
+        if (keyOrdering.lt(di.key, data.key)) balance(bNode(data, lsub.inc(di), rsub))
+        else if (keyOrdering.gt(di.key, data.key)) balance(bNode(data, lsub, rsub.inc(di)))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          bNode(d, lsub, rsub)
+        }
+      }
+  }
+}
+
+import tree._
+
+package infra {
+  class Inject[K, V](val keyOrdering: Ordering[K], val valueMonoid: Monoid[V])
+    extends Serializable {
+    def iNode(clr: Color, dat: Data[K], ls: Node[K], rs: Node[K]) =
+      new Inject[K, V](keyOrdering, valueMonoid) with INodeInc[K, V] with IncrementMap[K, V] {
+        // INode[K]
+        val color = clr
+        val lsub = ls.asInstanceOf[NodeInc[K, V]]
```

Comment:
`iNode` is an abstract function defined in `...redblack.tree.Node`  (the very top of the hierarchy).  Its return type is `INode` (the highest level internal tree node type).   At that level, it has no knowledge of any sub classes, and so `ls` and `rs` have to be `Node`, and therefore they need to be cast.   I also would have preferred not to require casting, but could not discover any way around it.  I decided it was at least (a) localized and (b) only relevant to anybody designing a new tree/map subclass.